### PR TITLE
Saml valve example project

### PR DESCRIPTION
**Pull request by ivy-arus**

### CODE REVIEW

File: SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import org.opensaml.xmlsec.signature.support.SignatureValidator;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+
+/**
+ * This class is responsible for authenticating a user given in a SAML Response
+ */
+public final class UserAuthenticator {
+  private static final String REQUEST_SAML_PARAMETER_NAME = "SAMLResponse";
+  private static final X509Certificate SIGN_VERIFICATION_CERT = loadCertificate("myCertificate");
```
Comment (alexsuter) [inline SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java:38]: how does this work? do I need "myCertifcate" file in my "user.dir"?

File: SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    // Add more Assertion validations here
+
+    enforceConditions(assertion.getConditions());
+  }
+
+  private static void validateSignature(Assertion assertion) throws AuthenticationException {
+    try {
+      BasicX509Credential credential = new BasicX509Credential(SIGN_VERIFICATION_CERT);
```
Comment (alexsuter) [inline SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java:134]: stick to a specific implementation?

File: SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  private static Assertion validateResponse(Response response) throws AuthenticationException {
+
+    String statusCode = response.getStatus().getStatusCode().getValue();
+    if (!statusCode.equals("urn:oasis:names:tc:SAML:2.0:status:Success")) {
+      throw new AuthenticationException("Invalid status code: " + statusCode);
+    }
+
+    if (response.getAssertions().size() != 1) {
```
Comment (alexsuter) [inline SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java:105]: hm. how does the assertion concept work? who is defining assertions? and what are these assertions? do you have somewhere a doc?

File: SamlValve/src/main/java/com/axonivy/demo/saml/UserConverter.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      return securityContext.executeAsSystemUser(callable);
+    } catch (Exception exc) {
+      throw new AuthenticationException("Unable to call findUser", exc);
+    }
+  }
+
+  private static String createIvyUser(ISecurityContext securityContext, String samlUserName, Assertion assertion,
+      Locale locale) throws AuthenticationException {
+    List<Attribute> attributes = assertion.getAttributeStatements().get(0).getAttributes();
+
```
Comment (alexsuter) [inline SamlValve/src/main/java/com/axonivy/demo/saml/UserConverter.java:59]: this should be configurable via tomcat valve attributes.    see our SSO valve in ivy-core.

File: SamlValve/src/main/java/com/axonivy/demo/saml/UserConverter.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    // Following attribute names are just examples.
+    String fullUserName = getAttribute(attributes, "fullName");
+    String eMailAddress = getAttribute(attributes, "eMail");
+    List<String> roles = getAttributeList(attributes, "roles");
+
```
Comment (alexsuter) [inline SamlValve/src/main/java/com/axonivy/demo/saml/UserConverter.java:64]: do we have somewhere doc about which common attributes which will be used. 
which attributes are for e.g. in fintech?

File: SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java (hunk 1 above)
Comment (ivy-arus) [reply SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java:38]: Currently, this points to the working dir. It is just thought as an example. Fintech specifies the certificate in a configuration file.

File: SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java (hunk 3 above)
Comment (ivy-arus) [reply SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java:105]: The assertion is part of the Identity Provider's Response, and this Response gets created after a user login.
1127182513244158 Xyqs8g4zWlwULTRe33zB_height640

File: SamlValve/src/main/java/com/axonivy/demo/saml/UserConverter.java (hunk 5 above)
Comment (ivy-arus) [reply SamlValve/src/main/java/com/axonivy/demo/saml/UserConverter.java:64]: No, we don't. I also googled it and couldn't find anything, and Emmanuel said that this is different for each implementation and product.
Fintech mapping can be found here:
https://bitbucket.org/axonivy-prod/ivy_tomcat_valve/src/master/ivy_tomcat_saml_valve/src/main/java/ch/axonivy/fintech/auth/valve/saml/assertion/SamlAssertionBuilder.java

File: SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java (hunk 2 above)
Comment (ivy-arus) [reply SamlValve/src/main/java/com/axonivy/demo/saml/UserAuthenticator.java]: Changed code

**Pull request closed without merging**